File: app/src/main/java/com/demos/kotlin/activity/charts/CombinedChartActivity.java
package com.demos.kotlin.activity.charts;

import android.graphics.Color;
import android.os.Bundle;
import android.view.WindowManager;

import com.demos.kotlin.R;
import com.demos.kotlin.activity.DemoBase;
import com.demos.kotlin.views.charts.CustomAxisValueFormatter;
import com.github.mikephil.charting.charts.CombinedChart;
import com.github.mikephil.charting.charts.CombinedChart.DrawOrder;
import com.github.mikephil.charting.components.Legend;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.components.XAxis.XAxisPosition;
import com.github.mikephil.charting.components.YAxis;
import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarDataSet;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.data.BubbleData;
import com.github.mikephil.charting.data.BubbleDataSet;
import com.github.mikephil.charting.data.BubbleEntry;
import com.github.mikephil.charting.data.CandleData;
import com.github.mikephil.charting.data.CandleDataSet;
import com.github.mikephil.charting.data.CandleEntry;
import com.github.mikephil.charting.data.CombinedData;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;
import com.github.mikephil.charting.data.ScatterData;
import com.github.mikephil.charting.data.ScatterDataSet;
import com.github.mikephil.charting.utils.ColorTemplate;

import java.util.ArrayList;

/**
 * 混合图
 */
public class CombinedChartActivity extends DemoBase {

    private CombinedChart chart;
    private final int count = 12;
    private int[] colors;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN);
        setContentView(R.layout.activity_combined);

        setTitle("CombinedChartActivity");
        colors = getResources().getIntArray(R.array.color_charts_detail);
        chart = findViewById(R.id.combined_chart);

        chart.getDescription().setEnabled(false);
        chart.setBackgroundColor(Color.WHITE);
        chart.setDrawBarShadow(false);
        chart.setHighlightFullBarEnabled(false);

        // draw bars behind lines
        chart.setDrawOrder(new DrawOrder[]{
                DrawOrder.BAR, DrawOrder.LINE
        });

        Legend l = chart.getLegend();
        l.setWordWrapEnabled(true);
        l.setVerticalAlignment(Legend.LegendVerticalAlignment.TOP);
        l.setHorizontalAlignment(Legend.LegendHorizontalAlignment.RIGHT);
        l.setOrientation(Legend.LegendOrientation.HORIZONTAL);
        l.setDrawInside(false);

        YAxis rightAxis = chart.getAxisRight();
        rightAxis.setDrawGridLines(true);
        rightAxis.setAxisMinimum(0f); // this replaces setStartAtZero(true)
        rightAxis.setAxisMaximum(100f); // this replaces setStartAtZero(true)
        rightAxis.setDrawAxisLine(false);

        YAxis leftAxis = chart.getAxisLeft();
        leftAxis.setDrawGridLines(false);
        leftAxis.setAxisMinimum(0f); // this replaces setStartAtZero(true)
        leftAxis.setAxisMaximum(100f); // this replaces setStartAtZero(true)
        leftAxis.setDrawAxisLine(false);

        XAxis xAxis = chart.getXAxis();
        xAxis.setPosition(XAxisPosition.BOTTOM);
        xAxis.setAxisMinimum(0f);
        xAxis.setDrawGridLines(false);
//        xAxis.setGranularity(1f);

        xAxis.setValueFormatter(new CustomAxisValueFormatter(CustomAxisValueFormatter.X_AXIS_MONTH));

        CombinedData data = new CombinedData();

        data.setData(generateLineData());
        data.setData(generateBarData());

        xAxis.setAxisMaximum(data.getXMax() + 0.25f);

        chart.setData(data);
        chart.invalidate();
    }

    /**
     * 折线图，初始化数据
     *
     * @return
     */
    private LineData generateLineData() {

        LineData d = new LineData();
        for (int i = 0; i < 3; i++) {
            ArrayList<Entry> entries = new ArrayList<>();
            for (int index = 0; index < count; index++)
                entries.add(new Entry(index + 0.5f, getRandom(100, 0)));
            LineDataSet set = new LineDataSet(entries, "Line DataSet " + i);
            set.setColor(colors[i]);
            set.setLineWidth(2.5f);
            set.setDrawCircles(false);
            set.setAxisDependency(YAxis.AxisDependency.RIGHT);
            d.addDataSet(set);
        }

        return d;
    }

    private BarData generateBarData() {

        ArrayList<BarEntry> entries1 = new ArrayList<>();
        ArrayList<BarEntry> entries2 = new ArrayList<>();
        ArrayList<BarEntry> entries3 = new ArrayList<>();

        for (int index = 0; index < count; index++) {
            entries1.add(new BarEntry(0, getRandom(100, 0)));
            entries3.add(new BarEntry(0, getRandom(100, 0)));

            // stacked
            entries2.add(new BarEntry(0, new float[]{getRandom(40, 0), getRandom(30, 0), getRandom(30, 0)}));
        }

        BarDataSet set1 = new BarDataSet(entries1, "Bar 1");
        set1.setColor(colors[0]);
        set1.setAxisDependency(YAxis.AxisDependency.LEFT);

        BarDataSet set2 = new BarDataSet(entries2, "");
        set2.setStackLabels(new String[]{"Stack 1", "Stack 2", "Stack 3"});
        set2.setColors(colors[1], colors[2], colors[3]);
        set2.setAxisDependency(YAxis.AxisDependency.LEFT);

        BarDataSet set3 = new BarDataSet(entries3, "Bar 3");
        set3.setColor(colors[4]);
        set3.setAxisDependency(YAxis.AxisDependency.LEFT);


        float groupSpace = 0.4f;
        float barSpace = 0.01f; // x2 dataset
        float barWidth = 0.19f; // x2 dataset
        // (0.45 + 0.02) * 2 + 0.06 = 1.00 -> interval per "group"

        BarData d = new BarData(set1, set2, set3);
        d.setBarWidth(barWidth);
        // make this BarData object grouped
        d.groupBars(0, groupSpace, barSpace); // start at x = 0

        return d;
    }

    private ScatterData generateScatterData() {

        ScatterData d = new ScatterData();

        ArrayList<Entry> entries = new ArrayList<>();

        for (float index = 0; index < count; index += 0.5f)
            entries.add(new Entry(index + 0.25f, getRandom(10, 55)));

        ScatterDataSet set = new ScatterDataSet(entries, "Scatter DataSet");
        set.setColors(ColorTemplate.MATERIAL_COLORS);
        set.setScatterShapeSize(7.5f);
        set.setDrawValues(false);
        set.setValueTextSize(10f);
        d.addDataSet(set);

        return d;
    }

    private CandleData generateCandleData() {

        CandleData d = new CandleData();

        ArrayList<CandleEntry> entries = new ArrayList<>();

        for (int index = 0; index < count; index += 2)
            entries.add(new CandleEntry(index + 1f, 90, 70, 85, 75f));

        CandleDataSet set = new CandleDataSet(entries, "Candle DataSet");
        set.setDecreasingColor(Color.rgb(142, 150, 175));
        set.setShadowColor(Color.DKGRAY);
        set.setBarSpace(0.3f);
        set.setValueTextSize(10f);
        set.setDrawValues(false);
        d.addDataSet(set);

        return d;
    }

    private BubbleData generateBubbleData() {

        BubbleData bd = new BubbleData();

        ArrayList<BubbleEntry> entries = new ArrayList<>();

        for (int index = 0; index < count; index++) {
            float y = getRandom(10, 105);
            float size = getRandom(100, 105);
            entries.add(new BubbleEntry(index + 0.5f, y, size));
        }

        BubbleDataSet set = new BubbleDataSet(entries, "Bubble DataSet");
        set.setColors(ColorTemplate.VORDIPLOM_COLORS);
        set.setValueTextSize(10f);
        set.setValueTextColor(Color.WHITE);
        set.setHighlightCircleWidth(1.5f);
        set.setDrawValues(true);
        bd.addDataSet(set);

        return bd;
    }

    @Override
    public void saveToGallery() { /* Intentionally left empty */ }
}

File: app/src/main/java/com/demos/kotlin/activity/kotlinchart/CombinedChartActivity.kt
package com.demos.kotlin.activity.kotlinchart

import android.graphics.Color
import android.os.Bundle
import android.view.WindowManager
import com.demos.kotlin.R
import com.demos.kotlin.activity.DemoBase
import com.demos.kotlin.databinding.ActivityCombinedBinding
import com.demos.kotlin.views.charts.CustomAxisValueFormatter
import com.github.mikephil.charting.charts.CombinedChart
import com.github.mikephil.charting.charts.CombinedChart.DrawOrder
import com.github.mikephil.charting.components.Legend
import com.github.mikephil.charting.components.XAxis.XAxisPosition
import com.github.mikephil.charting.components.YAxis
import com.github.mikephil.charting.data.*
import com.github.mikephil.charting.utils.ColorTemplate
import java.util.*

/**
 * 混合图
 */
class CombinedChartActivity : DemoBase() {
    private val count = 12
    private lateinit var colors: IntArray
    private lateinit var binding: ActivityCombinedBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN)
        binding = ActivityCombinedBinding.inflate(layoutInflater)
        setContentView(binding.root)
        title = "CombinedChartActivity"
        colors = resources.getIntArray(R.array.color_charts_detail)
        binding.combinedChart.getDescription().isEnabled = false
        binding.combinedChart.setBackgroundColor(Color.WHITE)
        binding.combinedChart.setDrawBarShadow(false)
        binding.combinedChart.setHighlightFullBarEnabled(false)

        // draw bars behind lines
        binding.combinedChart.setDrawOrder(arrayOf(
                DrawOrder.BAR, DrawOrder.LINE
        ))
        val l = binding.combinedChart.getLegend()
        l.isWordWrapEnabled = true
        l.verticalAlignment = Legend.LegendVerticalAlignment.TOP
        l.horizontalAlignment = Legend.LegendHorizontalAlignment.RIGHT
        l.orientation = Legend.LegendOrientation.HORIZONTAL
        l.setDrawInside(false)
        val rightAxis = binding.combinedChart.getAxisRight()
        rightAxis.setDrawGridLines(true)
        rightAxis.axisMinimum = 0f // this replaces setStartAtZero(true)
        rightAxis.axisMaximum = 100f // this replaces setStartAtZero(true)
        rightAxis.setDrawAxisLine(false)
        val leftAxis = binding.combinedChart.getAxisLeft()
        leftAxis.setDrawGridLines(false)
        leftAxis.axisMinimum = 0f // this replaces setStartAtZero(true)
        leftAxis.axisMaximum = 100f // this replaces setStartAtZero(true)
        leftAxis.setDrawAxisLine(false)
        val xAxis = binding.combinedChart.getXAxis()
        xAxis.position = XAxisPosition.BOTTOM
        xAxis.axisMinimum = 0f
        xAxis.setDrawGridLines(false)
        //        xAxis.setGranularity(1f);
        xAxis.valueFormatter = CustomAxisValueFormatter(CustomAxisValueFormatter.X_AXIS_MONTH)
        val data = CombinedData()
        data.setData(generateLineData())
        data.setData(generateBarData())
        xAxis.axisMaximum = data.xMax + 0.25f
        binding.combinedChart.setData(data)
        binding.combinedChart.invalidate()
    }

    /**
     * 折线图，初始化数据
     *
     * @return
     */
    private fun generateLineData(): LineData {
        val d = LineData()
        for (i in 0..2) {
            val entries = ArrayList<Entry>()
            for (index in 0 until count) entries.add(Entry(index + 0.5f, getRandom(100f, 0f)))
            val set = LineDataSet(entries, "Line DataSet $i")
            set.color = colors[i]
            set.lineWidth = 2.5f
            set.setDrawCircles(false)
            set.axisDependency = YAxis.AxisDependency.RIGHT
            d.addDataSet(set)
        }
        return d
    }

    private fun generateBarData(): BarData {
        val entries1 = ArrayList<BarEntry>()
        val entries2 = ArrayList<BarEntry>()
        val entries3 = ArrayList<BarEntry>()
        for (index in 0 until count) {
            entries1.add(BarEntry(0f, getRandom(100f, 0f)))
            entries3.add(BarEntry(0f, getRandom(100f, 0f)))

            // stacked
            entries2.add(BarEntry(0f, floatArrayOf(getRandom(40f, 0f), getRandom(30f, 0f), getRandom(30f, 0f))))
        }
        val set1 = BarDataSet(entries1, "Bar 1")
        set1.color = colors[0]
        set1.axisDependency = YAxis.AxisDependency.LEFT
        val set2 = BarDataSet(entries2, "")
        set2.stackLabels = arrayOf("Stack 1", "Stack 2", "Stack 3")
        set2.setColors(colors[1], colors[2], colors[3])
        set2.axisDependency = YAxis.AxisDependency.LEFT
        val set3 = BarDataSet(entries3, "Bar 3")
        set3.color = colors[4]
        set3.axisDependency = YAxis.AxisDependency.LEFT
        val groupSpace = 0.4f
        val barSpace = 0.01f // x2 dataset
        val barWidth = 0.19f // x2 dataset
        // (0.45 + 0.02) * 2 + 0.06 = 1.00 -> interval per "group"
        val d = BarData(set1, set2, set3)
        d.barWidth = barWidth
        // make this BarData object grouped
        d.groupBars(0f, groupSpace, barSpace) // start at x = 0
        return d
    }

    private fun generateScatterData(): ScatterData {
        val d = ScatterData()
        val entries = ArrayList<Entry>()
        var index = 0f
        while (index < count) {
            entries.add(Entry(index + 0.25f, getRandom(10f, 55f)))
            index += 0.5f
        }
        val set = ScatterDataSet(entries, "Scatter DataSet")
        set.setColors(*ColorTemplate.MATERIAL_COLORS)
        set.scatterShapeSize = 7.5f
        set.setDrawValues(false)
        set.valueTextSize = 10f
        d.addDataSet(set)
        return d
    }

    private fun generateCandleData(): CandleData {
        val d = CandleData()
        val entries = ArrayList<CandleEntry>()
        var index = 0
        while (index < count) {
            entries.add(CandleEntry(index + 1f, 90f, 70f, 85f, 75f))
            index += 2
        }
        val set = CandleDataSet(entries, "Candle DataSet")
        set.decreasingColor = Color.rgb(142, 150, 175)
        set.shadowColor = Color.DKGRAY
        set.barSpace = 0.3f
        set.valueTextSize = 10f
        set.setDrawValues(false)
        d.addDataSet(set)
        return d
    }

    private fun generateBubbleData(): BubbleData {
        val bd = BubbleData()
        val entries = ArrayList<BubbleEntry>()
        for (index in 0 until count) {
            val y = getRandom(10f, 105f)
            val size = getRandom(100f, 105f)
            entries.add(BubbleEntry(index + 0.5f, y, size))
        }
        val set = BubbleDataSet(entries, "Bubble DataSet")
        set.setColors(*ColorTemplate.VORDIPLOM_COLORS)
        set.valueTextSize = 10f
        set.valueTextColor = Color.WHITE
        set.highlightCircleWidth = 1.5f
        set.setDrawValues(true)
        bd.addDataSet(set)
        return bd
    }

    public override fun saveToGallery() { /* Intentionally left empty */
    }
}